File: 01_Language/01_Functions/java/src/main/java/com/onevgo/functions/Filetype.java
package com.onevgo.functions;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public enum Filetype {
    FIFO,
    CHAR,
    DIR,
    BLOCK,
    LINK,
    FILE,
    SOCKET,
    UNKNOWN;

    public static Filetype filetype(String filename) {
        try {
            Path path = Paths.get(filename);
            if (Files.isDirectory(path)) {
                return DIR;
            } else if (Files.isSymbolicLink(path)) {
                return LINK;
            } else if (Files.isRegularFile(path)) {
                return FILE;
            }
        } catch (RuntimeException e) {
            e.printStackTrace();
        }
        return UNKNOWN;
    }

    public static void main(String[] args) {
        System.out.println(filetype("/dev/urandom"));
        System.out.println(filetype("/dev/"));
        System.out.println(filetype("/dev/disk0"));
        System.out.println(filetype("/dev/stdin"));
        System.out.println(filetype("/etc/passwd"));
        System.out.println(filetype("/var/run/syslog"));
    }
}

File: 01_Language/01_Functions/go/src/functions/filetype.go
package functions

import "os"

func Filetype(filename string) string {
	if fileInfo, err := os.Lstat(filename); err == nil {
		mode := fileInfo.Mode()
		if mode&os.ModeNamedPipe != 0 {
			return "fifo"
		} else if mode&os.ModeCharDevice != 0 {
			return "char"
		} else if mode&os.ModeDevice != 0 {
			return "block"
		} else if mode.IsDir() {
			return "dir"
		} else if mode&os.ModeSymlink != 0 {
			return "link"
		} else if mode&os.ModeSocket != 0 {
			return "socket"
		} else if mode.IsRegular() {
			return "file"
		}
	}
	return "unknown"
}
